apps/dashboard/src/components/workflow-editor/steps/in-app/in-app-editor.tsx

```diff
@@ -0,0 +1,50 @@
+import { RiPencilRuler2Line } from 'react-icons/ri';
+import { UiSchemaGroupEnum, type UiSchema } from '@novu/shared';
+
+import { getComponentByType } from '@/components/workflow-editor/steps/component-utils';
+
+const avatarKey = 'avatar';
+const subjectKey = 'subject';
+const bodyKey = 'body';
+const redirectKey = 'redirect';
+const primaryActionKey = 'primaryAction';
+const secondaryActionKey = 'secondaryAction';
+
+export const InAppEditor = ({ uiSchema }: { uiSchema?: UiSchema }) => {
+  if (!uiSchema || uiSchema?.group !== UiSchemaGroupEnum.IN_APP) {
+    return null;
+  }
+
+  const {
+    [avatarKey]: avatar,
+    [subjectKey]: subject,
+    [bodyKey]: body,
+    [redirectKey]: redirect,
+    [primaryActionKey]: primaryAction,
+    [secondaryActionKey]: secondaryAction,
+  } = uiSchema.properties ?? {};
+
+  return (
+    <div className="flex flex-col gap-3">
+      <div className="flex items-center gap-2.5">
+        <RiPencilRuler2Line className="text-foreground-950 size-5 p-0.5 text-sm font-medium" />
+        <span>In-app Template</span>
+      </div>
+      <div className="flex flex-col gap-1 rounded-xl border border-neutral-100 p-1">
+        <div className="flex gap-1">
+          {avatar && getComponentByType({ component: avatar.component })}
+          {subject && getComponentByType({ component: subject.component })}
+        </div>
+        {body && getComponentByType({ component: body.component })}
+        {(primaryAction || secondaryAction) &&
+          getComponentByType({
```

Comment:
doesn't it means that if the secondary key is missing it will still show both options in the UI ? same for primary

Reply:
Hmmm, also when there is no action set yet, the widget should be displayed.

Reply:
> doesn't it means that if the secondary key is missing it will still show both options in the UI ? same for primary

No, this code is just responsible for rendering or not the "whole" component, but it has the logic inside that determines which buttons to render or not.

![Screenshot 2024-11-07 at 11 04 14](https://github.com/user-attachments/assets/ff1ba856-e702-46d1-9467-e05925cb291a)

Reply:
> Hmmm, also when there is no action set yet, the widget should be displayed.

`primaryAction || secondaryAction` variables are taken from the `controls.uiSchema`, which controls what components to render. But the `controls.dataSchema` default values or `controls.values` do control the "state" of the component, meaning that when there are no `control.values`, the component will still be rendered with a "no action" state.